File: ios/App/App/WebContentSurfacePlugin.swift
import UIKit
import WebKit
import Capacitor

@objc(WebContentSurfacePlugin)
public class WebContentSurfacePlugin: CAPPlugin, CAPBridgedPlugin, WKNavigationDelegate, WKScriptMessageHandler {
    public let identifier = "WebContentSurfacePlugin"
    public let jsName = "WebContentSurface"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "show", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hide", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setFrame", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "canGoBack", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "goBack", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reload", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "postMessage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setChatSoundEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "showCreateActions", returnType: CAPPluginReturnPromise)
    ]

    private var contentWebView: WKWebView?
    private var feedLoadingView: UIView?
    private var feedLoading = false
    private var feedLoadGeneration = 0
    private var chatSoundEnabled = true
    private var surfaceVisible = false
    private var allowedOrigin: URL?

    private func ensureWebView() -> WKWebView? {
        if let contentWebView { return contentWebView }
        guard let host = bridge?.viewController?.view else { return nil }

        let configuration = WKWebViewConfiguration()
        configuration.userContentController.add(self, name: "ChatPalezNativeSurface")
        configuration.websiteDataStore = .default()
        configuration.allowsInlineMediaPlayback = true
        configuration.mediaTypesRequiringUserActionForPlayback = []

        let webView = WKWebView(frame: .zero, configuration: configuration)
        // The dedicated WKWebView is outside Capacitor's primary web view, so
        // explicitly mirror the app UA marker required by mobile-session.php.
        webView.customUserAgent = "Mozilla/5.0 ChatPalezMobile/1.0"
        webView.backgroundColor = .white
        webView.isOpaque = true
        webView.isHidden = true
        webView.autoresizingMask = []
        webView.navigationDelegate = self
        host.addSubview(webView)
        host.bringSubviewToFront(webView)
        let loading = UIView(frame: .zero)
        loading.backgroundColor = .white
        loading.isHidden = true
        let spinner = UIActivityIndicatorView(style: .large)
        spinner.color = UIColor(red: 0, green: 102.0 / 255.0, blue: 178.0 / 255.0, alpha: 1)
        spinner.startAnimating()
        let label = UILabel()
        label.text = "Loading your feed…"
        label.textColor = UIColor(red: 18.0 / 255.0, green: 50.0 / 255.0, blue: 77.0 / 255.0, alpha: 1)
        label.font = .systemFont(ofSize: 16, weight: .medium)
        let stack = UIStackView(arrangedSubviews: [spinner, label])
        stack.axis = .vertical
        stack.alignment = .center
        stack.spacing = 16
        stack.translatesAutoresizingMaskIntoConstraints = false
        loading.addSubview(stack)
        NSLayoutConstraint.activate([
            stack.centerXAnchor.constraint(equalTo: loading.centerXAnchor),
            stack.centerYAnchor.constraint(equalTo: loading.centerYAnchor)
        ])
        host.addSubview(loading)
        feedLoadingView = loading
        contentWebView = webView
        return webView
    }

    @objc func open(_ call: CAPPluginCall) {
        guard let action = call.getString("action"),
              let token = call.getString("token"),
              let path = call.getString("path"),
              let requestedOrigin = call.getString("allowedOrigin"),
              let url = URL(string: action),
              let origin = URL(string: requestedOrigin),
              url.scheme?.lowercased() == "https",
              origin.scheme?.lowercased() == "https",
              url.host?.lowercased() == origin.host?.lowercased() else {
            call.reject("Missing or invalid authenticated web transition.")
            return
        }

        DispatchQueue.main.async {
            guard let webView = self.ensureWebView() else {
                call.reject("Unable to create ChatPalez web content surface.")
                return
            }

            self.allowedOrigin = origin
            self.applyFrame(call, to: webView)
            self.feedLoadGeneration += 1
            let generation = self.feedLoadGeneration
            self.feedLoading = path == "/"
            self.feedLoadingView?.isHidden = !self.feedLoading
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.setValue("application/x-www-form-urlencoded; charset=utf-8", forHTTPHeaderField: "Content-Type")
            request.httpBody = "token=\(self.formEncode(token))&path=\(self.formEncode(path))".data(using: .utf8)
            webView.isHidden = false
            self.surfaceVisible = true
            webView.superview?.bringSubviewToFront(webView)
            if self.feedLoading, let loading = self.feedLoadingView {
                loading.superview?.bringSubviewToFront(loading)
                DispatchQueue.main.asyncAfter(deadline: .now() + 15) { [weak self] in
                    guard let self, self.feedLoadGeneration == generation else { return }
                    self.finishFeedLoading()
                }
            }
            webView.load(request)
            call.resolve()
        }
    }

    @objc func show(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            if let webView = self.ensureWebView() {
                webView.isHidden = false
                self.surfaceVisible = true
                self.syncChatSound()
                webView.superview?.bringSubviewToFront(webView)
                if self.feedLoading, let loading = self.feedLoadingView {
                    loading.isHidden = false
                    loading.superview?.bringSubviewToFront(loading)
                }
            }
            call.resolve()
        }
    }

    @objc func hide(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.contentWebView?.isHidden = true
            self.surfaceVisible = false
            self.syncChatSound()
            self.feedLoadingView?.isHidden = true
            call.resolve()
        }
    }

    @objc func setFrame(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let webView = self.ensureWebView() else {
                call.reject("Unable to create ChatPalez web content surface.")
                return
            }
            self.applyFrame(call, to: webView)
            call.resolve()
        }
    }

    @objc func canGoBack(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve(["value": self.contentWebView?.canGoBack ?? false])
        }
    }

    @objc func goBack(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            if self.contentWebView?.canGoBack == true {
                self.contentWebView?.goBack()
            }
            call.resolve()
        }
    }

    @objc func showCreateActions(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let presenter = self.bridge?.viewController else {
                call.reject("Unable to present ChatPalez create actions.")
                return
            }

            let sheet = UIAlertController(
                title: "Create",
                message: "What would you like to share?",
                preferredStyle: .actionSheet
            )
            let actions: [(String, String)] = [
                ("Create Post", "post"),
                ("Upload Photos", "photos"),
                ("Create Story", "story"),
                ("Create Reel", "reel")
            ]
            for (title, value) in actions {
                sheet.addAction(UIAlertAction(title: title, style: .default) { _ in
                    call.resolve(["action": value])
                })
            }
            sheet.addAction(UIAlertAction(title: "Cancel", style: .cancel) { _ in
                call.resolve(["action": NSNull()])
            })

            // Keep the native iPhone action-sheet idiom while matching the
            // ChatPalez accent and supporting iPad's required popover anchor.
            sheet.view.tintColor = UIColor(red: 0, green: 102.0 / 255.0, blue: 178.0 / 255.0, alpha: 1)
            if let popover = sheet.popoverPresentationController {
                popover.sourceView = presenter.view
                popover.sourceRect = CGRect(
                    x: presenter.view.bounds.midX,
                    y: presenter.view.bounds.maxY - 1,
                    width: 1,
                    height: 1
                )
                popover.permittedArrowDirections = []
            }
            presenter.present(sheet, animated: true)
        }
    }

    @objc func postMessage(_ call: CAPPluginCall) {
        guard let message = call.getObject("message"),
              JSONSerialization.isValidJSONObject(message),
              let data = try? JSONSerialization.data(withJSONObject: message),
              let json = String(data: data, encoding: .utf8) else {
            call.reject("Missing or invalid web surface message.")
            return
        }
        DispatchQueue.main.async {
            guard let webView = self.contentWebView else {
                call.reject("ChatPalez web content surface is not open.")
                return
            }
            webView.evaluateJavaScript("window.dispatchEvent(new CustomEvent('chatpalez:native-surface-message',{detail:\(json)}));") { _, error in
                if let error {
                    call.reject("Unable to deliver web surface message.", nil, error)
                } else {
                    call.resolve()
                }
            }
        }
    }

    @objc func setChatSoundEnabled(_ call: CAPPluginCall) {
        guard let enabled = call.getBool("enabled") else {
            call.reject("Missing chat sound preference.")
            return
        }
        DispatchQueue.main.async {
            self.chatSoundEnabled = enabled
            self.syncChatSound()
            call.resolve()
        }
    }

    @objc func resetSession(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.contentWebView?.stopLoading()
            self.contentWebView?.loadHTMLString("", baseURL: nil)
            self.contentWebView?.isHidden = true
            self.surfaceVisible = false
            self.feedLoadGeneration += 1
            self.finishFeedLoading()
            self.allowedOrigin = nil
            let store = WKWebsiteDataStore.default()
            let types = WKWebsiteDataStore.allWebsiteDataTypes()
            store.fetchDataRecords(ofTypes: types) { records in
                store.removeData(ofTypes: types, for: records) {
                    call.resolve()
                }
            }
        }
    }

    @objc func reload(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.contentWebView?.reload()
            call.resolve()
        }
    }



    public func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        guard message.name == "ChatPalezNativeSurface",
              let body = message.body as? [String: Any],
              let type = body["type"] as? String,
              ["share", "pick-media", "open-native", "open-external"].contains(type) else { return }
        notifyListeners("command", data: body)
    }

    public func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        guard let url = navigationAction.request.url else {
            decisionHandler(.cancel)
            return
        }
        if isAllowed(url) {
            decisionHandler(.allow)
            return
        }
        if url.scheme?.lowercased() == "http" || url.scheme?.lowercased() == "https" {
            UIApplication.shared.open(url)
        }
        decisionHandler(.cancel)
    }

    public func webView(_ webView: WKWebView, decidePolicyFor navigationResponse: WKNavigationResponse, decisionHandler: @escaping (WKNavigationResponsePolicy) -> Void) {
        if let response = navigationResponse.response as? HTTPURLResponse, response.statusCode >= 400 {
            finishFeedLoading()
        }
        decisionHandler(.allow)
    }

    public func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
        notifyListeners("loadStarted", data: ["url": webView.url?.absoluteString ?? ""])
    }

    public func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        guard let url = webView.url, isAllowed(url) else { return }
        if url.path != "/mobile-session.php" { finishFeedLoading() }
        syncChatSound()
        notifyListeners("routeChanged", data: ["url": url.absoluteString])
        notifyListeners("loadFinished", data: ["url": url.absoluteString])
    }

    public func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        if (error as NSError).code != NSURLErrorCancelled { finishFeedLoading() }
    }

    public func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        if (error as NSError).code != NSURLErrorCancelled { finishFeedLoading() }
    }

    private func finishFeedLoading() {
        feedLoading = false
        feedLoadingView?.isHidden = true
    }

    private func syncChatSound() {
        guard let webView = contentWebView, allowedOrigin != nil else { return }
        let enabled = surfaceVisible && chatSoundEnabled ? "true" : "false"
        webView.evaluateJavaScript("if (typeof window.__chatpalezBaseChatSound === 'undefined') window.__chatpalezBaseChatSound = !!window.chat_sound; window.chat_sound = window.__chatpalezBaseChatSound && \(enabled);", completionHandler: nil)
    }

    private func isAllowed(_ url: URL) -> Bool {
        guard let trusted = allowedOrigin else { return false }
        return url.scheme?.lowercased() == "https" && url.host?.lowercased() == trusted.host?.lowercased()
    }

    private func applyFrame(_ call: CAPPluginCall, to webView: WKWebView) {
        let x = CGFloat(call.getDouble("x") ?? 0)
        let y = CGFloat(call.getDouble("y") ?? 0)
        let width = max(1, CGFloat(call.getDouble("width") ?? 1))
        let height = max(1, CGFloat(call.getDouble("height") ?? 1))
        webView.frame = CGRect(x: x, y: y, width: width, height: height)
        feedLoadingView?.frame = webView.frame
    }

    private func formEncode(_ value: String) -> String {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._*")
        return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? ""
    }
}

File: android/app/src/main/java/chatpalez/app/webview/WebContentSurfacePlugin.java
package chatpalez.app.webview;

import android.graphics.Color;
import android.net.Uri;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.CookieManager;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.webkit.JavascriptInterface;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceError;
import android.webkit.WebResourceResponse;
import android.content.Intent;
import android.widget.FrameLayout;
import android.app.Dialog;
import android.graphics.drawable.GradientDrawable;
import android.view.Gravity;
import android.view.Window;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.ProgressBar;
import android.widget.Button;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.AbsoluteSizeSpan;
import android.text.style.ForegroundColorSpan;
import android.text.style.StyleSpan;

import com.getcapacitor.JSObject;
import org.json.JSONObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;

@CapacitorPlugin(name = "WebContentSurface")
public class WebContentSurfacePlugin extends Plugin {
    private WebView contentWebView;
    private FrameLayout feedLoadingView;
    private boolean feedLoading;
    private int feedLoadGeneration;
    private boolean chatSoundEnabled = true;
    private boolean surfaceVisible;
    private String allowedOrigin;

    @Override
    public void load() {
        getActivity().runOnUiThread(this::ensureWebView);
    }

    private void ensureWebView() {
        if (contentWebView != null) return;

        ViewGroup host = getActivity().findViewById(android.R.id.content);
        if (host == null) return;

        contentWebView = new WebView(getContext());
        contentWebView.setBackgroundColor(Color.WHITE);
        contentWebView.setVisibility(View.GONE);

        WebSettings settings = contentWebView.getSettings();
        settings.setJavaScriptEnabled(true);
        // This secondary WebView does not inherit Capacitor's appendUserAgent.
        // Keep the backend's existing first-party mobile-session guard intact
        // by applying the same app marker used by the primary Capacitor WebView.
        String userAgent = settings.getUserAgentString();
        if (userAgent == null) userAgent = "";
        if (!userAgent.contains("ChatPalezMobile/1.0")) {
            settings.setUserAgentString(userAgent + " ChatPalezMobile/1.0");
        }
        settings.setDomStorageEnabled(true);
        settings.setDatabaseEnabled(true);
        settings.setMediaPlaybackRequiresUserGesture(false);
        settings.setAllowFileAccess(false);
        settings.setAllowContentAccess(true);

        CookieManager cookies = CookieManager.getInstance();
        cookies.setAcceptCookie(true);
        cookies.setAcceptThirdPartyCookies(contentWebView, true);

        contentWebView.addJavascriptInterface(new Object() {
            @JavascriptInterface
            public void postMessage(String raw) {
                getActivity().runOnUiThread(() -> emitCommand(raw));
            }
        }, "ChatPalezNativeSurface");

        contentWebView.setWebViewClient(new WebViewClient() {
            @Override
            public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
                if (request.isForMainFrame()) finishFeedLoading();
            }

            @Override
            public void onReceivedHttpError(WebView view, WebResourceRequest request, WebResourceResponse response) {
                if (request.isForMainFrame()) finishFeedLoading();
            }

            @Override
            public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
                Uri uri = request.getUrl();
                if (isAllowed(uri)) return false;
                if ("http".equalsIgnoreCase(uri.getScheme()) || "https".equalsIgnoreCase(uri.getScheme())) {
                    try {
                        getActivity().startActivity(new Intent(Intent.ACTION_VIEW, uri));
                    } catch (Exception ignored) {
                    }
                }
                return true;
            }

            @Override
            public void onPageStarted(WebView view, String url, android.graphics.Bitmap favicon) {
                JSObject data = new JSObject();
                data.put("url", url);
                notifyListeners("loadStarted", data);
            }

            @Override
            public void onPageFinished(WebView view, String url) {
                emitRouteChanged(url);
                Uri uri = Uri.parse(url);
                if (isAllowed(uri)) {
                    if (!"/mobile-session.php".equals(uri.getPath())) finishFeedLoading();
                    syncChatSound();
                    JSObject data = new JSObject();
                    data.put("url", url);
                    notifyListeners("loadFinished", data);
                }
            }
        });
        host.addView(contentWebView, new FrameLayout.LayoutParams(1, 1));
        contentWebView.bringToFront();

        feedLoadingView = new FrameLayout(getContext());
        feedLoadingView.setBackgroundColor(Color.WHITE);
        feedLoadingView.setVisibility(View.GONE);
        LinearLayout loadingContent = new LinearLayout(getContext());
        loadingContent.setOrientation(LinearLayout.VERTICAL);
        loadingContent.setGravity(Gravity.CENTER);
        int spacing = Math.round(16 * getContext().getResources().getDisplayMetrics().density);
        ProgressBar progress = new ProgressBar(getContext());
        loadingContent.addView(progress);
        TextView label = new TextView(getContext());
        label.setText("Loading your feed…");
        label.setTextColor(Color.rgb(18, 50, 77));
        label.setTextSize(16);
        LinearLayout.LayoutParams labelParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        labelParams.topMargin = spacing;
        loadingContent.addView(label, labelParams);
        feedLoadingView.addView(loadingContent, new FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT, Gravity.CENTER));
        host.addView(feedLoadingView, new FrameLayout.LayoutParams(1, 1));
    }

    @PluginMethod
    public void open(PluginCall call) {
        String action = call.getString("action");
        String token = call.getString("token");
        String path = call.getString("path");
        String requestedOrigin = call.getString("allowedOrigin");
        if (action == null || token == null || path == null || requestedOrigin == null) {
            call.reject("Missing authenticated web transition.");
            return;
        }

        Uri actionUri = Uri.parse(action);
        Uri originUri = Uri.parse(requestedOrigin);
        if (!"https".equalsIgnoreCase(actionUri.getScheme()) || !"https".equalsIgnoreCase(originUri.getScheme())
                || actionUri.getHost() == null || !actionUri.getHost().equalsIgnoreCase(originUri.getHost())) {
            call.reject("ChatPalez web content requires HTTPS.");
            return;
        }

        getActivity().runOnUiThread(() -> {
            ensureWebView();
            if (contentWebView == null) {
                call.reject("Unable to create ChatPalez web content surface.");
                return;
            }

            allowedOrigin = requestedOrigin;
            applyFrame(call);
            final int generation = ++feedLoadGeneration;
            feedLoading = "/".equals(path);
            feedLoadingView.setVisibility(feedLoading ? View.VISIBLE : View.GONE);
            String body = "token=" + encode(token) + "&path=" + encode(path);
            contentWebView.setVisibility(View.VISIBLE);
            surfaceVisible = true;
            contentWebView.bringToFront();
            if (feedLoading) feedLoadingView.bringToFront();
            contentWebView.postUrl(action, body.getBytes(StandardCharsets.UTF_8));
            if (feedLoading) contentWebView.postDelayed(() -> {
                if (feedLoadGeneration == generation) finishFeedLoading();
            }, 15000);
            call.resolve();
        });
    }

    @PluginMethod
    public void show(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            ensureWebView();
            if (contentWebView != null) {
                contentWebView.setVisibility(View.VISIBLE);
                surfaceVisible = true;
                syncChatSound();
                contentWebView.bringToFront();
                if (feedLoading) {
                    feedLoadingView.setVisibility(View.VISIBLE);
                    feedLoadingView.bringToFront();
                }
            }
            call.resolve();
        });
    }

    @PluginMethod
    public void hide(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            if (contentWebView != null) contentWebView.setVisibility(View.GONE);
            surfaceVisible = false;
            syncChatSound();
            if (feedLoadingView != null) feedLoadingView.setVisibility(View.GONE);
            call.resolve();
        });
    }

    @PluginMethod
    public void setFrame(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            ensureWebView();
            if (contentWebView == null) {
                call.reject("Unable to create ChatPalez web content surface.");
                return;
            }
            applyFrame(call);
            call.resolve();
        });
    }

    @PluginMethod
    public void canGoBack(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            JSObject result = new JSObject();
            result.put("value", contentWebView != null && contentWebView.canGoBack());
            call.resolve(result);
        });
    }

    @PluginMethod
    public void goBack(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            if (contentWebView != null && contentWebView.canGoBack()) contentWebView.goBack();
            call.resolve();
        });
    }

    @PluginMethod
    public void showCreateActions(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            final Dialog dialog = new Dialog(getActivity());
            dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);

            LinearLayout panel = new LinearLayout(getActivity());
            panel.setOrientation(LinearLayout.VERTICAL);
            int side = dp(18);
            panel.setPadding(side, dp(12), side, dp(16));

            GradientDrawable background = new GradientDrawable();
            background.setColor(Color.WHITE);
            background.setCornerRadii(new float[] { dp(24), dp(24), dp(24), dp(24), 0, 0, 0, 0 });
            panel.setBackground(background);

            TextView handle = new TextView(getActivity());
            GradientDrawable handleBackground = new GradientDrawable();
            handleBackground.setColor(Color.rgb(203, 213, 225));
            handleBackground.setCornerRadius(dp(3));
            handle.setBackground(handleBackground);
            LinearLayout.LayoutParams handleParams = new LinearLayout.LayoutParams(dp(42), dp(4));
            handleParams.gravity = Gravity.CENTER_HORIZONTAL;
            handleParams.bottomMargin = dp(12);
            panel.addView(handle, handleParams);

            TextView title = new TextView(getActivity());
            title.setText("Create");
            title.setTextSize(20);
            title.setTextColor(Color.rgb(17, 24, 39));
            title.setGravity(Gravity.CENTER);
            title.setTypeface(title.getTypeface(), android.graphics.Typeface.BOLD);
            LinearLayout.LayoutParams titleParams = new LinearLayout.LayoutParams(-1, -2);
            titleParams.bottomMargin = dp(10);
            panel.addView(title, titleParams);

            final String[] labels = { "Create post", "Upload photos", "Create story", "Create reel" };
            final String[] subtitles = { "Share an update with your community", "Choose photos from your device", "Post a moment to your story", "Create a short-form video" };
            final String[] actions = { "post", "photos", "story", "reel" };
            for (int i = 0; i < labels.length; i++) {
                final int index = i;
                Button action = new Button(getActivity());
                action.setAllCaps(false);
                String combined = labels[i] + "\n" + subtitles[i];
                SpannableString styledText = new SpannableString(combined);
                int labelEnd = labels[i].length();
                styledText.setSpan(new StyleSpan(android.graphics.Typeface.BOLD), 0, labelEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                styledText.setSpan(new AbsoluteSizeSpan(16, true), 0, labelEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                styledText.setSpan(new ForegroundColorSpan(Color.rgb(31, 41, 55)), 0, labelEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                styledText.setSpan(new AbsoluteSizeSpan(12, true), labelEnd + 1, combined.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                styledText.setSpan(new ForegroundColorSpan(Color.rgb(107, 114, 128)), labelEnd + 1, combined.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                action.setText(styledText);
                action.setTextColor(Color.rgb(31, 41, 55));
                action.setGravity(Gravity.START | Gravity.CENTER_VERTICAL);
                action.setPadding(dp(16), dp(8), dp(16), dp(8));
                GradientDrawable actionBg = new GradientDrawable();
                actionBg.setColor(Color.rgb(247, 250, 252));
                actionBg.setCornerRadius(dp(14));
                action.setBackground(actionBg);
                LinearLayout.LayoutParams actionParams = new LinearLayout.LayoutParams(-1, dp(62));
                actionParams.bottomMargin = dp(8);
                panel.addView(action, actionParams);
                action.setOnClickListener(v -> {
                    JSObject result = new JSObject();
                    result.put("action", actions[index]);
                    call.resolve(result);
                    dialog.dismiss();
                });
            }

            Button cancel = new Button(getActivity());
            cancel.setAllCaps(false);
            cancel.setText("Cancel");
            cancel.setTextSize(16);
            cancel.setTextColor(Color.rgb(0, 102, 178));
            GradientDrawable cancelBg = new GradientDrawable();
            cancelBg.setColor(Color.rgb(232, 243, 251));
            cancelBg.setCornerRadius(dp(14));
            cancel.setBackground(cancelBg);
            panel.addView(cancel, new LinearLayout.LayoutParams(-1, dp(50)));
            cancel.setOnClickListener(v -> {
                JSObject result = new JSObject();
                result.put("action", JSONObject.NULL);
                call.resolve(result);
                dialog.dismiss();
            });
            dialog.setOnCancelListener(ignored -> {
                JSObject result = new JSObject();
                result.put("action", JSONObject.NULL);
                call.resolve(result);
            });

            dialog.setContentView(panel);
            Window window = dialog.getWindow();
            if (window != null) {
                window.setBackgroundDrawableResource(android.R.color.transparent);
                window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
                window.setGravity(Gravity.BOTTOM);
                window.setDimAmount(0.34f);
                window.addFlags(android.view.WindowManager.LayoutParams.FLAG_DIM_BEHIND);
            }
            dialog.show();
            if (window != null) window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        });
    }

    private int dp(int value) {
        return Math.round(value * getContext().getResources().getDisplayMetrics().density);
    }

    @PluginMethod
    public void postMessage(PluginCall call) {
        JSObject message = call.getObject("message");
        if (message == null) {
            call.reject("Missing web surface message.");
            return;
        }
        getActivity().runOnUiThread(() -> {
            if (contentWebView == null) {
                call.reject("ChatPalez web content surface is not open.");
                return;
            }
            String json = JSONObject.quote(message.toString());
            contentWebView.evaluateJavascript(
                    "window.dispatchEvent(new CustomEvent('chatpalez:native-surface-message',{detail:JSON.parse(" + json + ")}));",
                    null
            );
            call.resolve();
        });
    }

    @PluginMethod
    public void setChatSoundEnabled(PluginCall call) {
        Boolean enabled = call.getBoolean("enabled");
        if (enabled == null) {
            call.reject("Missing chat sound preference.");
            return;
        }
        getActivity().runOnUiThread(() -> {
            chatSoundEnabled = enabled;
            syncChatSound();
            call.resolve();
        });
    }

    @PluginMethod
    public void resetSession(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            if (contentWebView != null) {
                contentWebView.stopLoading();
                contentWebView.loadUrl("about:blank");
                contentWebView.clearHistory();
                contentWebView.clearCache(false);
                contentWebView.setVisibility(View.GONE);
            }
            finishFeedLoading();
            ++feedLoadGeneration;
            surfaceVisible = false;
            CookieManager cookies = CookieManager.getInstance();
            cookies.removeAllCookies(value -> cookies.flush());
            allowedOrigin = null;
            call.resolve();
        });
    }

    @PluginMethod
    public void reload(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            if (contentWebView != null) contentWebView.reload();
            call.resolve();
        });
    }

    private void applyFrame(PluginCall call) {
        float density = getContext().getResources().getDisplayMetrics().density;
        int x = Math.round(call.getFloat("x", 0f) * density);
        int y = Math.round(call.getFloat("y", 0f) * density);
        int width = Math.max(1, Math.round(call.getFloat("width", 1f) * density));
        int height = Math.max(1, Math.round(call.getFloat("height", 1f) * density));

        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(width, height);
        params.leftMargin = x;
        params.topMargin = y;
        contentWebView.setLayoutParams(params);
        if (feedLoadingView != null) {
            FrameLayout.LayoutParams loadingParams = new FrameLayout.LayoutParams(width, height);
            loadingParams.leftMargin = x;
            loadingParams.topMargin = y;
            feedLoadingView.setLayoutParams(loadingParams);
        }
    }

    private void finishFeedLoading() {
        feedLoading = false;
        if (feedLoadingView != null) feedLoadingView.setVisibility(View.GONE);
    }

    private void syncChatSound() {
        if (contentWebView == null || allowedOrigin == null) return;
        contentWebView.evaluateJavascript(
                "if (typeof window.__chatpalezBaseChatSound === 'undefined') " +
                "window.__chatpalezBaseChatSound = !!window.chat_sound; " +
                "window.chat_sound = window.__chatpalezBaseChatSound && " +
                (surfaceVisible && chatSoundEnabled ? "true" : "false") + ";", null);
    }



    private void emitCommand(String raw) {
        try {
            JSObject data = JSObject.fromJSONObject(new JSONObject(raw));
            String type = data.getString("type");
            if (!"share".equals(type) && !"pick-media".equals(type)
                    && !"open-native".equals(type) && !"open-external".equals(type)) return;
            notifyListeners("command", data);
        } catch (Exception ignored) {
        }
    }

    private boolean isAllowed(Uri uri) {
        if (uri == null || allowedOrigin == null) return false;
        Uri trusted = Uri.parse(allowedOrigin);
        return "https".equalsIgnoreCase(uri.getScheme())
                && trusted.getHost() != null
                && trusted.getHost().equalsIgnoreCase(uri.getHost());
    }

    private void emitRouteChanged(String url) {
        Uri uri = Uri.parse(url);
        if (!isAllowed(uri)) return;
        JSObject data = new JSObject();
        data.put("url", url);
        notifyListeners("routeChanged", data);
    }

    private static String encode(String value) {
        try {
            return URLEncoder.encode(value, StandardCharsets.UTF_8.name());
        } catch (Exception error) {
            throw new IllegalStateException("Unable to encode authenticated web transition.", error);
        }
    }
}
